### DISCUSSION TITLE
0.13.0

### DISCUSSION THREAD

**Discussion by MohamedBassem** [karakeep-app/karakeep#127] [Announcements]
Welcome to the `0.13.0` release of Hoarder! This release is packed with a lot of new features to better organize your bookmarks. This includes: Nested lists, tag merging, a new suggested cleanups page, and more! Crawler also now takes screenshots and downloads the banner images locally. The [CLI](https://docs.hoarder.app/command-line) is now more powerful giving you ability to do mass modifications (e.g. archiving/adding to lists) until the mutli-select feature is developed.

> Note: You probably want to read the `Upgrading` section for that release. No breaking changes, but some nice recommendations are mentioned.

## Announcements

- The android app has been finally released on the play store ([link](https://play.google.com/store/apps/details?id=app.hoarder.hoardermobile&hl=en&gl=US))! Thanks a lot to all the beta testers during that time! Go check it out (and maybe give it a review while you're there 😉). The iOS app is still stuck in apple's review queue unfortunately, so you still need to use the [beta version](https://testflight.apple.com/join/QifpNxAT) for now.
- The new chrome/firefox extensions allow you to quickly add tags and lists to your newly saved bookmarks!

## New Features

- Nested lists are here! When you're creating a new list, you can choose a parent for that list and it'll be nested under it. The `All lists` page has been revamped to support showing the list tree.
- Merging tags is now a thing.
- A new `Cleanups` page that will suggest to you ways to cleanup your collection. As a start, it now detects potentially duplicate tags and suggests merging them. Currently, the algorithm is very simple and will detect duplicates if they have different cases (e.g. `Chess` and `chess`), and also tags that differ by at most one character (e.g. `web-devopment` and `Web development`). The inference job now is also fixed to avoid creating those duplicates in the first place.
  - In the future, tag duplication detection will be able to suggest merging semantically equivalent tags.
- The crawler now by default downloads the banner images. This is to protect against bookmarked links going away, but also to get around rate limits when you have multiple bookmarks from the same website (e.g. github). This will result into a larger storage space utilization. If you don't care about that, you can disable by setting `CRAWLER_DOWNLOAD_BANNER_IMAGE=false`.
- The crawler also will now take screenshots of the website it crawls and allows you to see them in the preview page. The screenshots will also be used as a fallback banner for links that don't have an image. Again, this will use some extra storage, so if you don't want screenshots, you can disable them with `CRAWLER_STORE_SCREENSHOT=false`.
- You can now edit lists to change its name or update its parent.
- You can now see (and manage) all the bookmarks a list is in from the new `Manage Bookmarks` drop down option.
- You can now delete tags from the `All tags` page. The `All tags` page will also show you the "unused" tags and gives you a quick way to mass delete them.
- The CLI got a lot of new commands. You can now update bookmarks from the CLI (e.g. to archive or favourite them), and you can add or remove bookmarks from lists. You can have a look on the updates docs page [here](https://docs.hoarder.app/command-line). Also, you can now get the CLI directly from NPM using `npm install -g @hoarderapp/cli`. The CLI is very useful for now to do mass actions on your bookmark while we develop the multi-select feature https://github.com/MohamedBassem/hoarder-app/issues/84.

## Fixes
- When you preview a bookmark (using the maximize button), the address bar of the browser will now show a unique link for that bookmark.
- The crawler now can correctly find the images in amazon links.
- The search indexing worker was in some rare cases incorrectly reporting success while it actually failed. This is now fixed.

### The new Cleanups page

<img width="259" alt="Screenshot 2024-04-26 at 2 19 02 PM" src="https://github.com/MohamedBassem/hoarder-app/assets/2418637/289efd35-34e2-4d2a-860e-b91774ae46be">

<img width="1071" alt="Screenshot 2024-04-26 at 2 17 28 PM" src="https://github.com/MohamedBassem/hoarder-app/assets/2418637/12d035e8-a4ee-4551-849c-26b5705c151a">

### Viewing and mass deleting unused tags

<img width="410" alt="Screenshot 2024-04-26 at 2 18 21 PM" src="https://github.com/MohamedBassem/hoarder-app/assets/2418637/08cb6b13-0b42-41af-9739-7571f38cfa67">

### Nested lists

<img width="1051" alt="Screenshot 2024-04-26 at 6 41 00 PM" src="https://github.com/MohamedBassem/hoarder-app/assets/2418637/d7bd9897-f1da-4343-87e8-d54c1a17fa41">

<img width="280" alt="Screenshot 2024-04-26 at 5 44 11 PM" src="https://github.com/MohamedBassem/hoarder-app/assets/2418637/06e642e2-3a4f-4917-87bd-00bdfb4273e0">

### The new extension capabilities
![image](https://github.com/MohamedBassem/hoarder-app/assets/2418637/ff58a630-b03f-4857-a4c5-5dd37759c9cc)

### Quick tag deletion

<img width="365" alt="Screenshot 2024-04-26 at 6 40 31 PM" src="https://github.com/MohamedBassem/hoarder-app/assets/2418637/b59fd2ce-5d62-4db8-900d-0d29e24df3df">

### Viewing the screenshots

<img width="1212" alt="Screenshot 2024-04-26 at 7 18 11 PM" src="https://github.com/MohamedBassem/hoarder-app/assets/2418637/4a8b2ec6-22c3-4518-8624-16278331c648">

## Upgrading

- If you're using HOARDER_VERSION=release, run docker compose pull && docker compose up -d.
- If you're pinning it to a specific version, upgrade the version and then run docker compose pull && docker compose up -d.

After Ugprading:
- Now that we're taking screenshots for pages, you'll want to tell chrome to hide scrollbars by adding `--hide-scrollbars` to the `command` section under the chrome container. Check the update docker compose ([here](https://github.com/MohamedBassem/hoarder-app/blob/main/docker/docker-compose.yml)) for reference on how to do it.
- Apparently, meillisearch collects analytics by default unless you opt out. Add `MEILI_NO_ANALYTICS=true` in the environment section of the meillisearch container if you don't want that. Check the update docker compose ([here](https://github.com/MohamedBassem/hoarder-app/blob/main/docker/docker-compose.yml)) for reference on how to do it. Thanks for the report and fix @vivekmiyani!
- The release introduced the ability to download banner images and take screenshots from webpages. Those will be applied only on newly crawled pages. if you want to do the same for your old bookmarks, you can re-crawl all the links from the admin panel by pressing on the `Recrawl All Links (Without Inference)`. Make sure you click the `(Without inference)` button otherwise you'll be re-running automatic tagging which can be expensive.

## All commits

- docker: Attempt to improve build times by changing cache backend by @MohamedBassem in #5d805e2
- fix(inference): Attempt to reuse existing identical tags by @MohamedBassem in #b4c7de2
- feature: Allow users to delete all unused tags in one go by @MohamedBassem in #7d163f2
- fix: Fix error when merging empty tags by @MohamedBassem in #0b02f94
- fix: Change tag listing API to return unused tags by @MohamedBassem in #5aabbec
- fix: Re-index affected bookmarks when one of their tags gets updated by @MohamedBassem in #4efcac7
- feature(crawler): Allow increasing crawler concurrency and configure storing images and screenshots by @MohamedBassem in #a91aff2
- feature: A new cleanups page to suggest ways to tidy up your bookmarks by @MohamedBassem in #9dace18
- feature(web): Add ability to rename, merge and fast delete tags. Fixes #105 (#125) by @MohamedBassem in #d07f2c9
- docs: Update docs to include the links to the android app by @MohamedBassem in #da6df7c
- fix(web): Fix hydration error in the bookmark preview page by @MohamedBassem in #92cff68
- ui(web): Show bookmark URL when maximizing a bookmark by @MohamedBassem in #ec15d20
- docs: Update the CLI docs to reference the published NPM package by @MohamedBassem in #4804f98
- style(cli): Change all help messages to lowercase by @MohamedBassem in #ac2f78f
- build(cli): Prepare for publishing CLI to npm by @MohamedBassem in #3352a3e
- feature(cli): Add ability to manipulate lists, tags and update bookmarks by @MohamedBassem in #af0cf9c
- fix(crawler): Better extraction for amazon images by @MohamedBassem in #5dac180
- fix(docs): Update the social image from the generic docusaurus image by @MohamedBassem in #ab92969
- feature(web): Change the AddList modal to allow adding and removing bookmarks from lists by @MohamedBassem in #88b92d8
- fix(workers): Increase robustness of search worker and add extra logging. Fixes #118 by @MohamedBassem in #c1baada
- fix(workers): Set a modern user agent and update the default viewport size by @MohamedBassem in #cca81ca
- release(extension): Release version 1.2.0 by @MohamedBassem in #3d1eeef
- feature(extension): Allow adding tags and lists to newly hoarded bookmarks by @MohamedBassem in #77b1aba
- fix(web): Fix missing link card image in list view by @MohamedBassem in #0e26095
- ui(extension): Use shadcn and better dark mode support by @MohamedBassem in #7ddcb5f
- fix(mobile): Add a bigger button to collable the nested list by @MohamedBassem in #5f599f2
- fix: Update docker-compose.yml (#119) by @Rosin1 in #cbdaf56
- docker: Disable meilisearch analytics (#117) by @vivekmiyani  in #9b9cdcf
- fix: Ensure that downloaded asset images are from the allowed content types by @MohamedBassem in #12c682b
- fix: Use next/image for serving bookmark images by @MohamedBassem in #e12fe02
- feature: Allow recrawling bookmarks without running inference jobs by @MohamedBassem in #4c589d4
- feature: Download images and screenshots by @MohamedBassem in #4402e6f
- mobile: Bump mobile version to 1.5 by @MohamedBassem in #b4a13ce
- fix(mobile): Fix the color of the collapsible triangle to respect dark mode by @MohamedBassem in #0c9a07e
- fix(mobile): Fix flicker on sign in page by @MohamedBassem in #4f9b6c8
- feature(mobile): Add support for nested listed on mobile by @MohamedBassem in #5ae6694
- feature: Nested lists (#110). Fixes #62 by @MohamedBassem in #e0999f7
- refactor: Move the shared types to the shared package by @MohamedBassem in #deba31e
- mobile: Bump mobile version to 1.4 by @MohamedBassem in #81fc4c4
- mobile: Add dark mode support by @MohamedBassem in #c46482c

**Comment by scubanarc** [comment]
Looking forward to testing this, but I can't pull it on unRAID.

`no matching manifest for linux/amd64 in the manifest list entries`

**Reply by MohamedBassem** [reply]
Ops. A last minute change resulted into images from different architectures overriding each other :) I fixed it now by fixing the `amd64` images. Will fix the `arm64` images next. Please try again and let me know how it goes.

**Reply by SeeJayEmm** [reply]
Came here to say the same thing.  This error is specifically for the Chrome container. I took a look at the manifest and I don't see a 123 tag anymore. I updated my compose to use 124 and that got it started.

**Reply by scubanarc** [reply]
That fixed it, thanks!

**Reply by MohamedBassem** [reply]
@SeeJayEmm version `123` of chrome seems to [exist](https://console.cloud.google.com/artifacts/docker/zenika-hub/us/gcr.io/alpine-chrome/sha256:e38563d4475a3d791e986500a2e4125c9afd13798067138881cf770b1f6f3980) and I was actually able to pull it. The problem was actually caused by `Hoarder`, I fixed that now and probably that's why it succeeded.

**Reply by MohamedBassem** [reply]
I pushed `0.13.1` to fix the ARM images as well. Sorry for the hassle.

**Comment by scubanarc** [comment]
Recrawl on Amazon product pages is working great.
